In [1]:
# Importing and installig packages
#pip install natsort
import numpy as np
import pandas as pd
from natsort import index_natsorted

In [2]:
# Uploading data from github
url = (r'https://gist.githubusercontent.com/bobbae/b4eec5b5cb0263e7e3e63a6806d045f2/raw/279b794a834a62dc108fc843a72c94c49361b501/data.csv')
data_original = pd.read_csv(url)
data_original.head()

Year  Rank           Company  Revenue (in millions) Profit (in millions)
0  1955     1    General Motors                 9823.5                  806
1  1955     2       Exxon Mobil                 5661.4                584.8
2  1955     3        U.S. Steel                 3250.4                195.4
3  1955     4  General Electric                 2959.1                212.6
4  1955     5            Esmark                 2510.8                 19.1

In [3]:
# Renaming the last two columns
data_original.rename(columns={'Revenue (in millions)':'Revenue', 'Profit (in millions)':'Profit'}, inplace=True)
data_original.columns

Index(['Year', 'Rank', 'Company', 'Revenue', 'Profit'], dtype='object')

In [4]:
# Counting the number of rows
row_count = len(data_original)
row_count

25500

## The results above shows that there are 25500 rows in the CSV data.


In [5]:
# Producing CSV file that contains the total row count for he CSV data
a = pd.Series(['25500'])
a.reset_index()
a.to_csv('row_count.csv',index = False, header=False)

In [6]:
# checking for data types and counting rows
data_original.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25500 entries, 0 to 25499
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Year     25500 non-null  int64  
 1   Rank     25500 non-null  int64  
 2   Company  25500 non-null  object 
 3   Revenue  25500 non-null  float64
 4   Profit   25500 non-null  object 
dtypes: float64(1), int64(2), object(2)
memory usage: 996.2+ KB


In [7]:
# Displacing the non-numeric rows in boolean format
df_profit_numeric = data_original['Profit'].apply(lambda x: not x.isnumeric())
df_profit_numeric.head(5)

0    False
1     True
2     True
3     True
4     True
Name: Profit, dtype: bool

In [8]:
# Making Profit a dataframe
df_profit_numeric =df_profit_numeric.to_frame()
df_profit_numeric.head()

Profit
0   False
1    True
2    True
3    True
4    True

In [9]:
# Renaming the column
df_profit_numeric.rename(columns={'Profit':'Profit_data_type'}, inplace=True)
df_profit_numeric.columns

Index(['Profit_data_type'], dtype='object')

In [10]:
# Counting the non-numeric rows in For Profit column
df_profit_numeric.value_counts()

Profit_data_type
True                20524
False                4976
dtype: int64

In [11]:
# Joining the profit column(boolean) and the original data
frames = [df_profit_numeric, data_original]
joined_df = pd.concat(frames, axis=1)
#joined_df.rename(columns={[0], 'Profit_data_type'}, inplace=True)
joined_df.head()

Profit_data_type  Year  Rank           Company  Revenue Profit
0             False  1955     1    General Motors   9823.5    806
1              True  1955     2       Exxon Mobil   5661.4  584.8
2              True  1955     3        U.S. Steel   3250.4  195.4
3              True  1955     4  General Electric   2959.1  212.6
4              True  1955     5            Esmark   2510.8   19.1

In [12]:
# Joining the df_profit_numeric data and the original data
frames = [df_profit_numeric, data_original]
joined_df = pd.concat(frames, axis=1)
joined_df.head()

Profit_data_type  Year  Rank           Company  Revenue Profit
0             False  1955     1    General Motors   9823.5    806
1              True  1955     2       Exxon Mobil   5661.4  584.8
2              True  1955     3        U.S. Steel   3250.4  195.4
3              True  1955     4  General Electric   2959.1  212.6
4              True  1955     5            Esmark   2510.8   19.1

In [13]:
# Deleting non-numeric rows for Profit column
df_clean = joined_df[joined_df.Profit_data_type != True]
df_clean.head()

Profit_data_type  Year  Rank                 Company  Revenue Profit
0              False  1955     1          General Motors   9823.5    806
19             False  1955    20                  Boeing   1033.2     37
31             False  1955    32                Uniroyal    782.6     28
38             False  1955    39     United Technologies    654.2     26
54             False  1955    55  Jones & Laughlin Steel    492.9     25

In [14]:
# Deleting the Profit_data_type column
df_clean.drop('Profit_data_type', axis=1, inplace=True)
df_clean.head()

/Users/nicolbiram/opt/anaconda3/lib/python3.7/site-packages/pandas/core/frame.py:4913: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  errors=errors,


Year  Rank                 Company  Revenue Profit
0   1955     1          General Motors   9823.5    806
19  1955    20                  Boeing   1033.2     37
31  1955    32                Uniroyal    782.6     28
38  1955    39     United Technologies    654.2     26
54  1955    55  Jones & Laughlin Steel    492.9     25

In [15]:
# reseting the unsorted index
df_clean.reset_index(drop=True, inplace=True)
df_clean.head()

Year  Rank                 Company  Revenue Profit
0  1955     1          General Motors   9823.5    806
1  1955    20                  Boeing   1033.2     37
2  1955    32                Uniroyal    782.6     28
3  1955    39     United Technologies    654.2     26
4  1955    55  Jones & Laughlin Steel    492.9     25

In [16]:
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4976 entries, 0 to 4975
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Year     4976 non-null   int64  
 1   Rank     4976 non-null   int64  
 2   Company  4976 non-null   object 
 3   Revenue  4976 non-null   float64
 4   Profit   4976 non-null   object 
dtypes: float64(1), int64(2), object(2)
memory usage: 194.5+ KB


## The results above shows that there are 4976 rows left in the data frame 

In [17]:
# Producing CSV data without non-numeric row for Profit 
df_clean.to_csv("numeric_rows.csv")

## Part 2

In [18]:
# counting rows for each column
df_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4976 entries, 0 to 4975
Data columns (total 5 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   Year     4976 non-null   int64  
 1   Rank     4976 non-null   int64  
 2   Company  4976 non-null   object 
 3   Revenue  4976 non-null   float64
 4   Profit   4976 non-null   object 
dtypes: float64(1), int64(2), object(2)
memory usage: 194.5+ KB


In [19]:
# turning the CSV file into a JSON file
df_clean.to_json("data2.json")

In [20]:
# Sorting the highest Profit
top_profit = df_clean.sort_values(by="Profit", key=lambda x: np.argsort(index_natsorted(df_clean["Profit"])) )

In [21]:
# Sorting and Printing the top 20 highest Profit
top20_profit = top_profit.tail(20)
top20_profit

Year  Rank                 Company   Revenue Profit
4792  2005    24                  Pfizer   52921.0  11361
4429  2003     3             Exxon Mobil  182466.0  11460
4067  2001    10  Verizon Communications   64707.0  11797
4062  2001     5        General Electric  129853.0  12735
4781  2005     6           ChevronTexaco  147967.0  13328
4063  2001     6               Citigroup  111826.0  13519
4254  2002     6        General Electric  125913.0  13684
4430  2003     5        General Electric  131698.0  14118
4255  2002     7               Citigroup  112022.0  14126
4790  2005    18   Bank of America Corp.   63324.0  14143
4591  2004     5        General Electric  134187.0  15002
4431  2003     6               Citigroup  100789.0  15276
4252  2002     2             Exxon Mobil  191581.0  15320
4780  2005     5        General Electric  152363.0  16593
4783  2005     8               Citigroup  108276.0  17046
4058  2001     1             Exxon Mobil  210392.0  17720
4594  2004     8               Citigroup   94713.0  17853
4588  2004     2             Exxon Mobil  213199.0  21510
3688  1999     2              Ford Motor  144416.0  22071
4777  2005     2             Exxon Mobil  270772.0  25330

### The above dataframe (top20_profit) shows the first 20 highest profit 

In [22]:
# Producing CSV data top 20 rows with the highest profit value
top20_profit.to_csv("top20_profit.csv")